File: strategy_pattern/python/1-duck-behavior/duck.py
from abc import abstractmethod


class Duck:

    def quack(self) -> None:
        print("Quack")

    def swim(self) -> None:
        print("Swim")

    @abstractmethod
    def display(self) -> None:
        pass

    def fly(self) -> None:
        print("I'm flying")


class MallardDuck(Duck):

    def display(self) -> None:
        print("I'm a Mallard Duck")


class RedheadDuck(Duck):

    def display(self) -> None:
        print("I'm a Redhead Duck")


class RubberDuck(Duck):

    def quack(self) -> None:
        print("Squeak")

    def display(self) -> None:
        print("I'm a Rubber Duck")

    def fly(self) -> None:
        print("I can't fly")


class DecoyDuck(Duck):

    def display(self) -> None:
        print("I'm a Decoy Duck")

    def fly(self) -> None:
        print("I can't fly")

    def quack(self) -> None:
        print("Silence")

File: strategy_pattern/java/1-duck-behavior/duck/Duck.java
package duck;


public abstract class Duck {

    FlyBehavior flyBehavior;
    QuackBehavior quackBehavior;

    public Duck() {}

    public void quack() {
        quackBehavior.quack();
    }

    public void swim() {
        System.out.println("Swim");
    }

    public abstract void display();

    public void fly() {
        flyBehavior.fly();
    }

    public void setFlyBehavior(FlyBehavior behavior) {
        flyBehavior = behavior;
    }

    public void setQuackBehavior(QuackBehavior behavior) {
        quackBehavior = behavior;
    }

}